ladybug_comfort/collection/pmv.py

```diff
@@ -0,0 +1,455 @@
+        self._hr_calculated = True
+
+    def _calculate_pmv(self):
+        """Compute PMV for each step of the Data Collection."""
+        if self._hr_comfort_required is True:
+            self._calculate_humidity_ratio()
+
+        # empty properties to be caulculated
+        self._pmv = []
+        self._ppd = []
+        self._set = []
+        self._is_comfortable = []
+        self._thermal_condition = []
+        self._discomfort_reason = []
+        self._ta_adj = []
+        self._cooling_effect = []
+        self._heat_loss_conduction = []
+        self._heat_loss_sweating = []
+        self._heat_loss_latent_respiration = []
+        self._heat_loss_dry_respiration = []
+        self._heat_loss_radiation = []
+        self._heat_loss_convection = []
+
+        # perform the PMV calculation
+        for ta, tr, vel, rh, met, clo, wme, i in \
+            zip(self._air_temperature, self._rad_temperature,
+                self._air_speed, self._rel_humidity,
+                self._met_rate, self._clo_value,
+                self._external_work, range(self._calc_length)):
+            result = predicted_mean_vote(ta, tr, vel, rh, met, clo, wme,
+                                         self._comfort_par.still_air_threshold)
+            self._pmv.append(result['pmv'])
+            self._ppd.append(result['ppd'])
+            self._set.append(result['set'])
+            self._ta_adj.append(result['ta_adj'])
+            self._cooling_effect.append(result['ce'])
+            self._heat_loss_conduction.append(result['heat_loss']['cond'])
+            self._heat_loss_sweating.append(result['heat_loss']['sweat'])
+            self._heat_loss_latent_respiration.append(result['heat_loss']['res_l'])
+            self._heat_loss_dry_respiration.append(result['heat_loss']['res_s'])
+            self._heat_loss_radiation.append(result['heat_loss']['rad'])
+            self._heat_loss_convection.append(result['heat_loss']['conv'])
+
+            # determine whether conditions are acceptable
+            condit = self._comfort_par.thermal_condition(
+                result['pmv'], result['ppd'])
+            if self._hr_comfort_required is True:
+                hr = self._humidity_ratio[i]
+                comf = self._comfort_par.is_comfortable(result['ppd'], hr)
+                reason = self._comfort_par.discomfort_reason(
+                    result['pmv'], result['ppd'], hr)
+            else:
+                comf = self._comfort_par.is_comfortable(result['ppd'])
+                reason = self._comfort_par.discomfort_reason(
+                    result['pmv'], result['ppd'])
+            self._is_comfortable.append(comf)
+            self._thermal_condition.append(condit)
+            self._discomfort_reason.append(reason)
+
+    @property
+    def air_temperature(self):
+        """Data Collection of air temperature values in degrees C."""
+        return self._air_temperature.duplicate()
```

Comment:
This seemed to be the best way that I could make the input properties to the comfort model objects immutable.  If I do not duplicate the data collections here, someone could edit the input air temperature values on this data collection and then the air temperature values would no longer correspond to the calculated PMV values on the comfort model object.  The alternative to this is having the PMV recalculate every time that someone edits a value on the air temperature data collection but implementing this would be a nightmare and would not be computationally efficient at all.  The way it is now, someone can always get the air temperature data collection that was input to the comfort model, do whatever they want to it and the PMV object will still have the correct inputs/outputs on it.  It may also be worth noting that the cases are rare when people need to get the input data collections from the comfort model object are rare since they will usually have the inputs whenever they create the comfort model object.

Reply:
There are ways to do this with introducing less overhead. The most common way is to overwrite `__setattr__` . You can even get fancy and write a decorator to take care of that. See `frozen` decorator in honeybee as an example: https://github.com/ladybug-tools/honeybee/blob/a05669780d6846ff6574cd3eed55eb40e39c9e22/honeybee/radiance/parameters/_frozen.py

You can also just overwrite the `__setattr__` in this class and ensure the properties are protected.

But it seems this is something that can happen in other cases so we might implement this to DataCollection class itself. Then one can set it to `read_only` after creation.

Reply:
@mostaphaRoudsari ,
I see what you mean and I agree that this is a situation that will be common enough to merit a read_only attribute on the DataCollection object.  Thank you also for the example with the frozen decorator as this gives me a good place to start and I will begin working on a PR to implement this on the DataCollection.

My only potential misgiving about going down this route is that it will mean that there is a check to see if the DataCollection is `read_only `every time that someone goes to edit it.  This could decrease the efficiency of workflows like creating the DataCollection object and then setting values individually since this `read_only `check will be called 8760 times for an annual hourly data collection.  I could see this inefficiency as a relatively small price to pay for the convenience of being easily able to set the data collection to read only.  However, this means that I will be more likely to use workflows where I create a list of 8760 values and then set it to the DataCollection all at once since this will only involve a single check for the read_only attribute instead of 8760 individual ones.  I remember you thought it was odd that I did this in a recent PR so I just want to make sure that you think this practice is ok since it will be all over the source code if we implement a `read_only `check like this on the DataCollection.

Reply:
What about creating an immutable instance of datacollection. Similar to tuple versus list? This will be easier to implement and all one needs to do in similar cases is to create a immutable instance of the datacollection. It is also a more explicit solution.

Reply:
The more that I am thinking about this, the more I realize that `read_only `collections do not seem like the right solution for the comfort collection objects.  Or rather, I think it is a situation of an edge case driving the core development of many workflows.  I feel this way because, even if we implemented `read_only` collections, I would still have to duplicate the original data collections that are input to the comfort model in order to avoid changing those original collections to `read_only` without the user's permission.  The only difference is that I would have to do this duplication every time that someone makes a comfort collection object instead of only doing in when someone needs to get the original inputs to the comfort collection, which (let's face it) is an edge case since people usually have the original input collection when they create the comfort collection object.

I think a better way to handle this is similar to how I currently handle the outputs of the comfort model, where the results are stored in hidden lists on the object after the computation of the comfort model happens upon initialization.  These lists are only used to create Data Collections when someone requests a given output and this avoids both the creation of extra data collection objects for outputs that the user does not care about and it avoids the possibility of an output collection editing the original comfort collection object since the lists on the main object are hidden.

To me, this seems like a lot simpler and more elegant a solution than adding a `read_only` attribute to Data Collections.  I also worry about how a `read_only` attribute on the data collection could affect larger workflows where collections are being passed through a series of operations but, if we agree that there is no immediate need for this now, we can have this conversation later if need be.

Reply:
Sounds good to me!

Reply:
Glad that we are in agreement on this.  I made the changes accordingly.